Fetch full `developer.log` messages

elliette opened the pull request:
Fixes  https://github.com/dart-lang/webdev/issues/2330, https://github.com/flutter/devtools/issues/3222

The CDP [consoleAPICalled](https://chromedevtools.github.io/devtools-protocol/tot/Runtime/#event-consoleAPICalled) event args are [`RemoteObjects`](https://chromedevtools.github.io/devtools-protocol/tot/Runtime/#type-RemoteObject) with only the `objectPreview` provided. This `objectPreview` is what we were previously constructing the `RemoteLog` from. Since they are only a preview, the log messages are abbreviated.

To get the full log message, we need to fetch the full object using [`getProperties`](https://chromedevtools.github.io/devtools-protocol/tot/Runtime/#method-getProperties).

Confirmed that we now see unabbreviated logs in DevTools:

<img width="1564" alt="Screenshot 2024-01-09 at 2 42 29 PM" src="https://github.com/dart-lang/webdev/assets/21270878/59f04554-26e7-4999-a363-f2ac1f62922b">

elliette commented:
Will update the CHANGELOG once I reset DWDS (it was just released)

nshahan commented on dwds/lib/src/services/chrome_proxy_service.dart:
```diff
@@ -1402,12 +1408,13 @@ ${globalToolConfiguration.loadStrategy.loadModuleSnippet}("dart_sdk").developer.
-        logParams[obj['name'] as String] = RemoteObject(obj);
-      }
-    }
+    final objectId = logObject?["objectId"];
```
nit: use single quotes to match the surrounding context.

elliette replied:
Done!

nshahan reviewed (approved):
I'm so excited you found the answer to the truncating message! Thank you so much!

Is there an easy way to reproduce the truncating so that we can test it is no longer happening?

And just to document our in person conversation, we can look into the `inspect()` API as well.

elliette commented:
> I'm so excited you found the answer to the truncating message! Thank you so much!
> 
> Is there an easy way to reproduce the truncating so that we can test it is no longer happening?
> 
> And just to document our in person conversation, we can look into the `inspect()` API as well.

Added a test case, and opened a bug and added a TODO for inspect

annagrin commented on dwds/test/chrome_proxy_service_test.dart:
```diff
@@ -2370,25 +2370,43 @@ void main() {
+      test('long dart:developer log messages are not truncated', () async {
+        final logStream = context.service.onEvent(EventStreams.kLogging);
+        final longMessage =
+            'A very long log message that Chrome truncates by default and requires users to expand in order to see the entire message.';
```
Nit: the long line, can change it to

'A very long log message that Chrome truncates by default'
' and requires users to expand in order to see the entire message.'

elliette replied:
Done!

annagrin commented on dwds/lib/src/services/chrome_proxy_service.dart:
```diff
@@ -1436,6 +1441,28 @@ ${globalToolConfiguration.loadStrategy.loadModuleSnippet}("dart_sdk").developer.
 
+  Future<Map<String, RemoteObject>> _fetchFullLogParams(String objectId) async {
+    final logParams = <String, RemoteObject>{};
+    for (final property in await inspector.getProperties(objectId)) {
```
I wonder if the inspector is always ready when the logs are printed - maybe switch to the abbreviated version if the inspector is not ready or fails to retrieve properties?

elliette replied:
Done!

toineheuvelmans commented:
My `webdev` version is 3.4.0, which [should](https://pub.dev/packages/dwds/changelog) include this fix, but I still see truncated logs. Am I missing something?

elliette commented:
Hi @toineheuvelmans ! Could you file an issue here with steps to reproduce? https://github.com/dart-lang/webdev/issues/new?assignees=&labels=package%3Awebdev&projects=&template=webdev.md

It's possible we aren't able to fetch the full logs here: https://github.com/dart-lang/webdev/blob/2024775a15dc3089cd9eca5dc55513a1311e41d1/dwds/lib/src/services/chrome_proxy_service.dart#L1486

Thanks!

toineheuvelmans commented:
Hi @elliette, here you go, I hope this suffices: https://github.com/dart-lang/webdev/issues/2400
